pkg/ddl/rollingback.go

```diff
@@ -662,6 +624,11 @@ func convertJob2RollbackJob(w *worker, jobCtx *jobContext, job *model.Job) (ver
 		err = dbterror.ErrCancelledDDLJob
 	}
 
+	if job.State == model.JobStateRollingback {
+		// try the best effort to rollback, ignore the cancellation.
+		jobCtx.ctx = w.workCtx
```

Comment:
harder to reason if we change context in the middle of execution, please add more comment for this field

the `context` we want to cancel if the job transit to `canceling` state should be the context of the step, not the context of job execution loop, we still need to handle `cancelling -> rollback if needed -> cancelled` step, see ideas below.

other ideas I can think of now is
- have a context per step, only start the cancel check routine on non-cancel/done/rollbackdone state, but it might be too heavy for general DDL, they normally runs fast, we might need to test the real overhead.
- only start the check routine for reorg jobs, but then this loop is a little duplicate with transitOneJobStepAndWaitSync loop, 1 should be enough

Reply:
I want to discuss this topic "the context belongs to job step or job or other things?" here @tangenta 

My idea is the job workflow has two directions: forward and rollingback. Both directions have multiple job steps. Only the forward direction can cancel the context. And my original thought is there are two types of job context like `jobContext` and `rollbackJobContext`, only the former one has a `cancel` member. But I think we don't bother to add a new struct. Maybe I can add a `IsRollback` to `jobContext` and add `getCtx` to return `jobCtx.ctx` or `context.Background()` based on it?

Another idea is the `getCtx` function should have a `JobState` argument, and check state to return `jobCtx.ctx` or `context.Background()`.

> have a context per step

I think this is similar to "`getCtx` function should have a `JobState` argument" and has less overhead.

> only start the check routine for reorg jobs

For simplicity I don't want to treat reorg jobs specially.

Reply:
the thing we want to cancel is the running step when job state changed by other session, and the job worker loop should keeps running other following steps.

in this pr we create a context for the entire job worker loop, and then we change it if we detect it's cancelled, to keep the code working, quite anti-pattern. I expect something similar to this: 
if we need to cancel a subtask, then we cancel its own context, leave upper layers of context unchanged
https://github.com/pingcap/tidb/blob/025c2a3c5b72557b1a4fc2df7c0ee7be8f551957/pkg/disttask/framework/taskexecutor/task_executor.go#L384-L404

Reply:
> the thing we want to cancel is the running step when job state changed by other session, and the job worker loop should keeps running other following steps.

This does not describe the property "rollback steps should not be canceled" or "cancelation only happens once". Extra logic should be introduced to accomplish this step context model and make it more complex. I think the two direction model is more clear.

To avoid replace the context on the fly, maybe this code is more clear

```golang
func transitOneJobStep() {
  if job.IsRollingback() {
    worker.runOneRollingbackJobStep(worker.ctx, jobCtx, job)
  } else {
    worker.runOneJobStep(jobCtx.ctx, jobCtx, job)
  }
}
```

But it's a bit weird that the `jobCtx` has `Ctx` in its name but doesn't have a `ctx` member.

Or we construct another `jobCtx`?

```golang
jobStepCtx := jobCtx
if job.IsRollingback() {
  jobStepCtx = newRollingbackJobCtx(jobCtx)
}
worker.runOneJobStep(jobStepCtx, job)
```

Reply:
@D3Hunter We cannot cancel the own context directly in `admin cancel` because it may be executed in another TiDB. In any case, another loop is needed, whether it is put to the "entire job worker" or "step".

Reply:
> @D3Hunter We cannot cancel the own context directly in `admin cancel` because it may be executed in another TiDB. In any case, another loop is needed, whether it is put to the "entire job worker" or "step".

yes, we have consensus on that. I think we are discussing context lifecycle and avoiding changing context here

Reply:
updated https://github.com/pingcap/tidb/pull/56404/commits/3516e3c36765fcd59c06e16b49f9a6c6599e1868